pkg/controllers/trace_controller_test.go

```diff
@@ -0,0 +1,208 @@
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/types"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/api/v1alpha1"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+)
+
+// FakeGadget is a fake implementation of the Gadget interface for tests. It records
+// the calls to its methods for assertions in the unit tests
+type FakeGadget struct {
+	mu    sync.Mutex
+	calls map[string]struct{}
+}
+
+func (g *FakeGadget) Delete(trace types.NamespacedName) error {
+	g.mu.Lock()
+	g.calls["delete/"+trace.String()] = struct{}{}
+	g.mu.Unlock()
+	return nil
+}
+
+func (g *FakeGadget) Operation(trace *gadgetv1alpha1.Trace, operation string, params map[string]string) error {
+	g.mu.Lock()
+	key := fmt.Sprintf("operation/%s/%s/%s/",
+		trace.ObjectMeta.Namespace,
+		trace.ObjectMeta.Name,
+		operation,
+	)
+	paramsKeys := make([]string, 0, len(params))
+	for k := range params {
+		paramsKeys = append(paramsKeys, k)
+	}
+	sort.Strings(paramsKeys)
+
+	for _, k := range paramsKeys {
+		key += fmt.Sprintf("%s/%s/", k, params[k])
+	}
+	g.calls[key] = struct{}{}
+	g.mu.Unlock()
+
+	trace.Status.OperationError = "FakeError"
+	trace.Status.State = "FakeState"
+	trace.Status.Output = "FakeOutput"
+	return nil
+}
+
+// methodHasBeenCalled is a helper function to check if a method has been
+// called on the gadget
+func methodHasBeenCalled(fakeGadget *FakeGadget, key string) func() string {
+	return func() string {
+		fakeGadget.mu.Lock()
+		defer fakeGadget.mu.Unlock()
+		_, ok := fakeGadget.calls[key]
+		if ok {
+			delete(fakeGadget.calls, key)
+			return key
+		} else {
+			return fmt.Sprint(fakeGadget.calls)
+		}
+	}
+}
+
+// OperationMethodHasBeenCalled returns a Gomega assertion checking if the
+// method Operation() has been called
+func OperationMethodHasBeenCalled(fakeGadget *FakeGadget, name, operation, params string) func() bool {
+	key := fmt.Sprintf("operation/%s/%s/%s",
+		name,
+		operation,
+		params,
+	)
+	return func() bool {
+		f := methodHasBeenCalled(fakeGadget, key)
+		return f() == key
+	}
+}
+
+// DeleteMethodHasBeenCalled returns a Gomega assertion checking if the method
+// Delete() has been called
+func DeleteMethodHasBeenCalled(fakeGadget *FakeGadget, name string) func() bool {
+	key := "delete/" + name
+	return func() bool {
+		f := methodHasBeenCalled(fakeGadget, key)
+		return f() == key
+	}
+}
+
+// UpdatedTrace returns a function that fetches the Trace object in a way that
+// can be used in Gomega's 'Eventually' or 'Consistently' methods.
+func UpdatedTrace(ctx context.Context, key client.ObjectKey) func() *gadgetv1alpha1.Trace {
+	trace := &gadgetv1alpha1.Trace{}
+
+	return func() *gadgetv1alpha1.Trace {
+		err := k8sClient.Get(ctx, key, trace)
+		if err != nil {
+			return nil
```

Comment:
Is there any explicit check to avoid a nil de-reference if this is returned?

Reply:
You're right, it crashes in `HaveState()` and similar functions if `Get()` return nil. I've fixed `HaveState()` and similar functions, and added a test that `trace` should be nil before the resource is created.